File: apps/worker/src/worker/asset_processing/image_pipeline.py
import io
import logging
import uuid
from datetime import datetime, timezone

from bson import ObjectId
from PIL import Image

from worker.libs.cloudflare import download_image_from_r2
from worker.libs.mongodb import get_knowledge_base_images_collection
from worker.libs.openai.embeddings import embed_texts
from worker.libs.openai.vision import describe_image_with_vision
from worker.libs.pinecone import get_image_index_name, get_or_create_index

log = logging.getLogger(__name__)


def _trace_id() -> str:
    return str(uuid.uuid4())[:8]


def _get_page_for_image(image_id: str) -> int:
    """Fetch pageNumber from KnowledgeBaseImages."""
    coll = get_knowledge_base_images_collection()
    doc = coll.find_one({"_id": ObjectId(image_id)}, {"pageNumber": 1})
    return doc.get("pageNumber", 0) if doc else 0


def _process_single_image(payload: dict) -> None:
    """Step 1: Update status. Step 2: Download. Step 3: Vision. Step 4: Embed. Step 5: Pinecone. Step 6: Update DB."""
    trace = _trace_id()
    image_id = payload["imageId"]
    kb_id = payload["knowledgeBaseId"]
    r2_key = payload["r2Key"]

    log.info("[trace_id=%s] Processing image imageId=%s", trace, image_id)

    # Step 1: Update status to INDEXING
    coll = get_knowledge_base_images_collection()
    coll.update_one(
        {"_id": ObjectId(image_id)},
        {"$set": {"indexingStatus": "INDEXING", "updatedAt": datetime.now(timezone.utc)}},
    )

    # Step 2: Download from R2
    img_bytes = download_image_from_r2(r2_key)
    pil_img = Image.open(io.BytesIO(img_bytes)).convert("RGB")

    # Step 3: Vision model
    description = describe_image_with_vision(pil_img)

    # Step 4: Embed
    [embedding] = embed_texts([description])

    # Step 5: Upsert to Pinecone
    index = get_or_create_index(get_image_index_name())
    vec_id = f"{kb_id}::img::{image_id}"
    page_num = _get_page_for_image(image_id)
    index.upsert(
        vectors=[
            {
                "id": vec_id,
                "values": embedding,
                "metadata": {
                    "kb_id": kb_id,
                    "page": page_num,
                    "text": description[:1000],
                    "image_id": image_id,
                },
            }
        ]
    )

    # Step 6: Update DB
    get_knowledge_base_images_collection().update_one(
        {"_id": ObjectId(image_id)},
        {
            "$set": {
                "textSummary": description,
                "embeddingId": vec_id,
                "indexingStatus": "INDEXED",
                "updatedAt": datetime.now(timezone.utc),
            }
        },
    )
    log.info("[trace_id=%s] Image indexed: %s", trace, image_id)


def run_image_pipeline(payload: dict) -> dict:
    _process_single_image(payload)
    return {"imageId": payload["imageId"], "status": "INDEXED"}

File: apps/worker/src/asset-processing/image-pipeline.ts
import { randomUUID } from "crypto";
import { PrismaClient } from "@repo/database";
import { downloadImageFromR2 } from "../libs/cloudflare.js";
import { embedTexts } from "../libs/openai/embeddings.js";
import { describeImageWithVision } from "../libs/openai/vision.js";
import { getImageIndex } from "../libs/pinecone.js";
import { getSettings } from "../core/config.js";

const prisma = new PrismaClient();

function traceId(): string {
  return randomUUID().slice(0, 8);
}

export interface ImagePayload {
  imageId: string;
  knowledgeBaseId: string;
  r2Key: string;
  pageNumber: number;
}

export async function runImagePipeline(payload: ImagePayload): Promise<{
  imageId: string;
  status: string;
}> {
  const results = await runImagePipelineBatch([payload]);
  return results[0];
}

export async function runImagePipelineBatch(
  payloads: ImagePayload[]
): Promise<Array<{ imageId: string; status: string }>> {
  if (payloads.length === 0) return [];

  const trace = traceId();
  const s = getSettings();
  const batchSize = s.PINECONE_UPSERT_BATCH_SIZE;

  console.info(
    `[trace_id=${trace}] Processing image batch: ${payloads.length} image(s)`
  );

  const imageIds = payloads.map((p) => p.imageId);
  await prisma.knowledgeBaseImages.updateMany({
    where: { id: { in: imageIds } },
    data: { indexingStatus: "INDEXING" },
  });

  let imgBuffers: Buffer[];
  try {
    imgBuffers = await Promise.all(
      payloads.map((p) => downloadImageFromR2(p.r2Key))
    );
  } catch (err) {
    console.error(`[trace_id=${trace}] Failed to download images from R2:`, err);
    await prisma.knowledgeBaseImages.updateMany({
      where: { id: { in: imageIds } },
      data: {
        indexingStatus: "ERROR",
        errorMessage: (err as Error).message,
      },
    });
    throw err;
  }

  const descriptions = await Promise.all(
    imgBuffers.map((buf) => describeImageWithVision(buf))
  );

  let embeddings: number[][];
  try {
    embeddings = await embedTexts(descriptions);
  } catch (err) {
    console.error(`[trace_id=${trace}] Failed to generate embeddings:`, err);
    await prisma.knowledgeBaseImages.updateMany({
      where: { id: { in: imageIds } },
      data: {
        indexingStatus: "ERROR",
        errorMessage: (err as Error).message,
      },
    });
    throw err;
  }

  const index = await getImageIndex();
  const vectors = payloads.map((p, i) => ({
    id: `${p.knowledgeBaseId}::img::${p.imageId}`,
    values: embeddings[i],
    metadata: {
      kb_id: p.knowledgeBaseId,
      page: p.pageNumber,
      text: descriptions[i].slice(0, 1000),
      image_id: p.imageId,
    },
  }));

  try {
    for (let i = 0; i < vectors.length; i += batchSize) {
      const batch = vectors.slice(i, i + batchSize);
      await index.upsert(batch);
    }
  } catch (err) {
    console.error(
      `[trace_id=${trace}] Failed to upsert image embeddings to Pinecone:`,
      err
    );
    await prisma.knowledgeBaseImages.updateMany({
      where: { id: { in: imageIds } },
      data: {
        indexingStatus: "ERROR",
        errorMessage: (err as Error).message,
      },
    });
    throw err;
  }

  console.info(
    `[trace_id=${trace}] Image embeddings upserted to Pinecone: ${vectors.length} vectors`
  );

  const vecIds = vectors.map((v) => v.id);
  const kbIdToVecIds = new Map<string, string[]>();
  for (let i = 0; i < payloads.length; i++) {
    const kbId = payloads[i].knowledgeBaseId;
    const vecId = vecIds[i];
    const list = kbIdToVecIds.get(kbId) ?? [];
    list.push(vecId);
    kbIdToVecIds.set(kbId, list);
  }

  await Promise.all(
    payloads.map((p, i) =>
      prisma.knowledgeBaseImages.update({
        where: { id: p.imageId },
        data: {
          textSummary: descriptions[i],
          embeddingId: vecIds[i],
          indexingStatus: "INDEXED",
        },
      })
    )
  );

  for (const [kbId, ids] of kbIdToVecIds) {
    if (ids.length === 1) {
      const kb = await prisma.knowledgeBase.findUnique({
        where: { id: kbId },
        select: { imageEmbeddingIds: true },
      });
      const current = kb?.imageEmbeddingIds ?? [];
      await prisma.knowledgeBase.update({
        where: { id: kbId },
        data: { imageEmbeddingIds: [...current, ids[0]] },
      });
    } else {
      const kb = await prisma.knowledgeBase.findUnique({
        where: { id: kbId },
        select: { imageEmbeddingIds: true },
      });
      const current = kb?.imageEmbeddingIds ?? [];
      await prisma.knowledgeBase.update({
        where: { id: kbId },
        data: { imageEmbeddingIds: [...current, ...ids] },
      });
    }
  }

  console.info(
    `[trace_id=${trace}] Image batch indexed: ${payloads.length} image(s)`
  );

  return payloads.map((p) => ({ imageId: p.imageId, status: "INDEXED" }));
}
